dist/tools/zep_dispatch/topology.c

```diff
@@ -0,0 +1,272 @@
+/*
+ * Copyright (C) 2021 Benjamin Valentin
+ *
+ * This file is subject to the terms and conditions of the GNU General Public
+ * License v2. See the file LICENSE for more details.
+ */
+
+#include <arpa/inet.h>
+#include <netdb.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <limits.h>
+
+#include "kernel_defines.h"
+#include "topology.h"
+#include "zep_parser.h"
+
+#define NODE_NAME_MAX_LEN   32
+#define HW_ADDR_MAX_LEN      8
+
+struct node {
+    list_node_t next;
+    char name[NODE_NAME_MAX_LEN];
+    uint8_t mac[HW_ADDR_MAX_LEN];
+    struct sockaddr_in6 addr;
+    uint8_t mac_len;
+};
+
+struct edge {
+    list_node_t next;
+    struct node *a;
+    struct node *b;
+    float weight_a_b;
+    float weight_b_a;
+};
+
+static char *_fmt_addr(char *out, size_t out_len, const uint8_t *addr, uint8_t addr_len)
+{
+    char *start = out;
+
+    if (out_len < 3 * addr_len) {
+        return NULL;
+    }
+
+    while (addr_len--) {
+        out += sprintf(out, "%02X", *addr++);
+        *(out++) = addr_len ? ':' : '\0';
+    }
+
+    return start;
+}
+
+static struct node *_find_node_by_name(const list_node_t *nodes, const char *name)
+{
+    for (list_node_t *node = nodes->next; node; node = node->next) {
+        struct node *super = container_of(node, struct node, next);
+        if (strncmp(super->name, name, sizeof(super->name)) == 0) {
+            return super;
+        }
+    }
+
+    return NULL;
+}
+
+static struct node *_find_or_create_node(list_node_t *nodes, const char *name)
+{
+    struct node *node = _find_node_by_name(nodes, name);
+
+    if (node == NULL) {
+        node = malloc(sizeof(*node));
+        strncpy(node->name, name, sizeof(node->name) - 1);
+        node->mac_len = 0;
+        list_add(nodes, &node->next);
+    }
+
+    return node;
+}
+
+static bool _parse_line(char *line, list_node_t *nodes, list_node_t *edges)
+{
+    struct edge *e;
+
+    if (*line == '#') {
+        return true;
+    }
+
+    char *a     = strtok(line, "\t ");
+    char *b     = strtok(NULL, "\n\t ");
+    char *e_ab  = strtok(NULL, "\n\t ");
+    char *e_ba  = strtok(NULL, "\n\t ");
+
+    if (a == NULL || b == NULL) {
+        return false;
+    }
+
+    if (e_ab == NULL) {
+        e_ab = "1";
+    }
+
+    if (e_ba == NULL) {
+        e_ba = e_ab;
+    }
+
+    e = malloc(sizeof(*e));
+
+    e->a = _find_or_create_node(nodes, a);
+    e->b = _find_or_create_node(nodes, b);
+    e->weight_a_b = atof(e_ab);
+    e->weight_b_a = atof(e_ba);
+
+    list_add(edges, &e->next);
+
+    return true;
+}
+
+int topology_print(const char *file, const topology_t *t)
+{
+    FILE *out;
+    char addr_str[3 * HW_ADDR_MAX_LEN];
+
+    if (t->flat) {
+        // TODO
```

Comment:
?

Reply:
Since printing flat typologies is rather pointless, I did not implement it.